### PR TITLE
boards: add support for the Waveshare nRF52840 Eval Kit

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR adds the support for the [Waveshare nRF52840 Eval Kit](https://www.waveshare.com/wiki/NRF52840_Eval_Kit).

The main on-board components of the Waveshare nRF52840 Eval Kit are:
- Arduino headers for connecting Arduino shields
- Raspberry Pi GPIO header for connecting Raspberry Pi HATs
- USB to UART bridge
- Optical sensor (LDR)
- 2 user keys RESET and KEY
- 4 User LEDs
- TF card slot
- CR2032 battery holder

The board definition include SAUL support as well Arduino support including Arduino PWM. The follwing peripherals and on-board components were tested:

- [x] ADC tested with `tests/periph_adc` / `tests/sys_arduino_analog`
- [x] GPIO tested with `tests/periph_gpio`
- [x] PWM tested with `tests/periph_pwm` / `tests/sys_arduino_analog`
- [x] UART tested with  `tests/periph_uart`
- [x] I2C tested with `tests/driver_sht3x`
- [x] SPI tested with `tests/driver_bme680`
- [x] SD card tested with `tests/driver_sdcard_spi`
- [x] Arduino compatibility tested with `tests/sys_arduino_analog`
- [x] Bluetooth connectivity
- [x] IEEE 802.15.4 connectivity tested with `examples/gnrc_networking` and driver `nrf802154`

[UPDATE] ~Commit https://github.com/RIOT-OS/RIOT/commit/84db9c269736e9a33d3a8c49489130dc583ccc0f can be dropped once PR #13858 is merged.~

### Issues/PRs references

~Depends on PR #14480~

### CODE REVIEW

**Comment by miri64** [comment]
@RIOT-OS/maintainers does anyone has this board at hand for testing?

**Comment by MrKevinWeiss** [comment]
We just got some in.  I'll have a look when I am back in the office.

**Comment by benpicco** [comment]
Probably unrelated, but when I'm flashing my [`nRF52810` based board](https://de.aliexpress.com/item/4000867497654.html) with ST-Link I will always get

> nRF52 device has a CTRL-AP dedicated to recover the device from AP lock.
> A high level adapter (like a ST-Link) you are currently using cannot access
> the CTRL-AP so 'nrf52_recover' command will not work.
> Do not enable UICR APPROTECT.

When I use my jlink adapter I can flash it just fine. I thought that protection was from the original firmware and would be gone after first flashing RIOT, but I keep getting that message with ST-Link.

**Comment by benpicco** [comment]
This board does not come with a programmer, so I don't see why it should depend on #14480.
`common/nrf52/Makefile.include` will default to JLink, I'm sure the board can be flashed with that.

Having the possibility to flash it with ST-Link is of course nice, but that's not requirement.

Btw, please squash.

**Comment by gschorcht** [comment]
> This board does not come with a programmer, so I don't see why it should depend on #14480.
> `common/nrf52/Makefile.include` will default to JLink, I'm sure the board can be flashed with that.

Of course, you are right.

**Comment by benpicco** [comment]
Oh just two things I noticed when adding support for a nRF52-module based board you should check:

 - [ ] `tests/periph_rtt` works
 - [ ] `examples/nimble_scanner` does not get stuck on init (and if you have the e.g. Corona Warn App installed, you should see periodic announcements with `scan`)

If that works, all clocks should be set up all-right. (my module didn't have any external oscillators)

File: boards/waveshare-nrf52840-eval-kit/board.c
Comment (benpicco) [inline boards/waveshare-nrf52840-eval-kit/board.c:37]: no longer needed :wink:

File: boards/waveshare-nrf52840-eval-kit/include/board.h
Comment (benpicco) [inline boards/waveshare-nrf52840-eval-kit/include/board.h:83]: Want to add a 

```C
#if defined(MODULE_MTD_SDCARD)
 /* this is provided by the sdcard_spi driver
  * see drivers/sdcard_spi/sdcard_spi.c */
extern sdcard_spi_t sdcard_spi_devs[ARRAY_SIZE(sdcard_spi_params)];
mtd_sdcard_t mtd_sdcard_dev = {
    .base = {
        .driver = &mtd_sdcard_driver,
    },
    .sd_card = &sdcard_spi_devs[0],
    .params = &sdcard_spi_params[0]
};

mtd_dev_t *mtd0 = (mtd_dev_t *)&mtd_sdcard_dev;

#if defined(MODULE_FATFS_VFS)
#include "fs/fatfs.h"
VFS_AUTO_MOUNT(fatfs, VFS_MTD(mtd_sdcard_dev), VFS_DEFAULT_SD(0), 0);
#endif
#endif /* MODULE_MTD_SDCARD || DOXYGEN */
```

to `board.c` or `mtd.c` and 

```
ifneq (,$(filter mtd,$(USEMODULE)))
  USEMODULE += mtd_sdcard_spi
endif

# default to using littlefs2 on the external flash and fatfs on SD card
ifneq (,$(filter vfs_default,$(USEMODULE)))
  USEMODULE += fatfs_vfs
  USEMODULE += mtd
endif
``` 

to `Makefile.dep` - then `tests/vfs_default`, `tests/mtd_raw` should just work.

File: boards/waveshare-nrf52840-eval-kit/Makefile.dep
Comment (benpicco) [inline boards/waveshare-nrf52840-eval-kit/Makefile.dep:9]:
```suggestion
# default to using fatfs on SD card
```

**Review by benpicco — APPROVED** [review]
Please squash

File: boards/waveshare-nrf52840-eval-kit/Makefile.dep
Comment (gschorcht) [reply boards/waveshare-nrf52840-eval-kit/Makefile.dep:9]: I just took your suggested code :wink:

**Comment by gschorcht** [comment]
At some point in the future, we should solve Doxygen's thousands-warning problem due to self-explanatory undocumented macros in board definitions. It is absolutely annoying to have to include all warnings in the exclude_patterns file. The file already has a size of more than 2 MByte.

**Comment by gschorcht** [comment]
By using regular expressions the number of new entries for the new board in `exclude_pattern` could be reduced considerably. But it is still annoying.

**Comment by gschorcht** [comment]
Static tests are now successful. May I still squash?

**Comment by benpicco** [comment]
Squash as much as you want :smile:

**Comment by gschorcht** [comment]
Hm, there is still a problem with Kconfig modelling and makefile dependencies. Probably, we need the fix in PR #18395.

**Comment by gschorcht** [comment]
@benpicco Kconfig seems to work now, see the small fixes. Commit https://github.com/RIOT-OS/RIOT/pull/13860/commits/23c20fbfdbc789435de1c289cb63b6f2d6984283 is the fix in PR #18395. Once PR #18395, this PR has to be rebased. May squash the fixes when rebasing?

**Comment by benpicco** [comment]
No need to ask for approval to squash in such tiny changes

**Comment by MrKevinWeiss** [comment]
I would also be OK skipping the compile job after this one is complete as it is the intermittent unrelated failure.

**Comment by MrKevinWeiss** [comment]
Though the binary size difference is a bit scary.

**Comment by gschorcht** [comment]
> I would also be OK skipping the compile job after this one is complete as it is the intermittent unrelated failure.

Oops, too late. I aborted and restarted the compilation test already, also to give your Murdock test the precedence :wink:

**Comment by gschorcht** [comment]
After 3 tries of compilation tests that failed each time due to the unrelated hash mismatch problem (this time in [compile/tests/saul/esp32-wrover-kit:gnu](https://ci.riot-os.org/RIOT-OS/RIOT/13860/8475804b5a9509beb9cd768d07efed363b9e85cf/output/compile/tests/saul/esp32-wrover-kit:gnu.txt)), I would restart the CI test without compilation test.

**Pull request merged**
